lib/auth/init.go

```diff
@@ -985,3 +992,67 @@ func migrateRoleOptions(ctx context.Context, asrv *Server) error {
 
 	return nil
 }
+
+// DELETE IN: 7.0.0
+// migrateMFADevices migrates registered MFA devices to the new storage format.
+func migrateMFADevices(ctx context.Context, asrv *Server) error {
+	users, err := asrv.GetUsers(true)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	for _, user := range users {
+		la := user.GetLocalAuth()
+		if la == nil {
+			continue
+		}
+		if len(la.MFA) > 0 {
+			// User already migrated.
+			continue
+		}
+
+		if len(la.TOTPKey) > 0 {
+			d, err := types.NewTOTPDevice("totp", la.TOTPKey, asrv.clock.Now())
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			la.MFA = append(la.MFA, d)
+
+			la.TOTPKey = ""
+		}
+		if la.U2FRegistration != nil {
+			pubKeyI, err := x509.ParsePKIXPublicKey(la.U2FRegistration.PubKey)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			pubKey, ok := pubKeyI.(*ecdsa.PublicKey)
+			if !ok {
+				return trace.BadParameter("expected *ecdsa.PublicKey, got %T", pubKeyI)
+			}
+			d, err := types.NewU2FDevice("u2f", &u2f.Registration{
+				KeyHandle: la.U2FRegistration.KeyHandle,
+				PubKey:    *pubKey,
+			}, asrv.clock.Now())
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			d.GetU2F().Counter = la.U2FCounter
+			la.MFA = append(la.MFA, d)
+
+			la.U2FRegistration = nil
+			la.U2FCounter = 0
+		}
+
+		if len(la.MFA) == 0 {
+			// No MFA devices to migrate.
+			continue
+		}
+
+		user.SetLocalAuth(la)
+		if err := asrv.UpsertUser(user); err != nil {
```

Comment:
Nit: I would consider adding a debug log entry that the user's local auth is being migrated.